### PR TITLE
DDPB-4341 allow us to recover from disasters in automated fashion

### PR DESCRIPTION
**Pull request by jamesrwarren**
## Purpose

Restore from snapshots and point in time as well as from backup account.
Fixes DDPB-4341

## Approach
The snapshot method is fairly annoying and obtuse and works differently based on serverless and clusters. Given the size of the database I would be half tempted to just use pg_dumps but snapshots are more AWS-ey and I had already got far with the snapshots by the time I realised how annoying it was. Also I'm not sure how to easily do PITR with pg_dumps method.

Anyway there are some examples and explanations of how it works in the readme. It's quite slow because of all the renaming, deleting, restoring etc... but not horrific but still slow.

## Learning
Some of the defaults aren't good so learned we had to use more settings from the original or terraform tries to recreate DB!

## Checklist
- [x] I have performed a self-review of my own code
- [x] I have updated documentation (Confluence/ADR/tech debt doc) where relevant
- [ ] I have added tests to prove my work
- [ ] The product team have approved these changes

### Frontend
- [ ] I have run an in-browser accessibility test (e.g. WAVE, Lighthouse)
- [ ] There are no deprecated CSS classes noted in the profiler
- [ ] Translations are used and the profiler doesn't identify any missing
- [ ] Any links or buttons added are screen reader friendly and contextually complete
- [ ] If adding GA events, I have updated or [checked](docs/runbooks/GOOGLE-ANALYTICS.md) the existing category or label values

### CODE REVIEW

File: disaster_recovery/restore/disaster_recovery.py [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+        if self.db_cluster_identifier_source is not None:
+            self.get_cluster_info()
+            if self.EngineMode != 'serverless':
+                self.get_instance_info()
+            if self.db_cluster_identifier_source == self.db_cluster_identifier_target and self.DeletionProtection:
+                self.drop_protection()
+
+        self.apply_overrides()
+
+        if self.PointInTimeRecovery != datetime.strptime('1900-01-01 00:00:00', '%Y-%m-%d %H:%M:%S'):
```
Comment (acsauk) [inline disaster_recovery/restore/disaster_recovery.py:102]: Assuming this is the format `self.PointInTimeRecovery` must be set to in order to restore to a point in time we could add a little more context here by extracting `'1900-01-01 00:00:00'` to a var with a name describing its intent (something like `requiredDateTimeFormat`?)

File: disaster_recovery/restore/disaster_recovery.py [hunk 2]
```diff
@@ -0,0 +1,621 @@
...
+            self.db_cluster_identifier_target = f"{self.db_cluster_identifier_source}-temp"
+
+        self.restore_cluster_ss()
+        if self.EngineMode != 'serverless':
+            self.create_db_instances()
+
+        if self.same_target:
+            self.overwrite_existing_cluster()
+
+    def restore_cluster_ss(self):
```
Comment (acsauk) [inline disaster_recovery/restore/disaster_recovery.py:180]: What does `ss` stand for? Could we get it expanded to the full words?

File: disaster_recovery/restore/disaster_recovery.py [hunk 3]
```diff
@@ -0,0 +1,621 @@
...
+    def restore_to_point_in_time(self):
+        if self.same_target:
+            self.db_cluster_identifier_target = f"{self.db_cluster_identifier_source}-temp"
+
+        self.restore_cluster_pitr()
+
+        if self.same_target:
+            self.overwrite_existing_cluster()
+
+    def restore_cluster_pitr(self):
```
Comment (acsauk) [inline disaster_recovery/restore/disaster_recovery.py:224]: Same as `ss` - could we get this expanded to `point_in_time_xxxx` (not sure what the r stands for 😅 )

File: docs/DISASTER_RECOVERY.md [hunk 4]
```diff
@@ -23,11 +23,65 @@ All of our measures to restore the service fall within the 48 hours specified by
...
+Title: Restore to a different cluster
+Example disaster: You have no idea when something bad happened to the data and want to go back in time on 
+another instance and see if the data was ok at a point in time without affecting your main DB operation.
+Remediation: Restore from an existing DB to a point in time into another cluster.
+```
+aws-vault exec identity --duration=2h -- \
+docker-compose -f docker-compose.commands.yml run -rm dr-restore \
+python3 disaster_recovery.py \
+--cluster_from api-ddpb9999 --cluster_to api-new_test_cluster --pitr '2022-01-01 09:10:00' 
+```
```
Comment (acsauk) [inline docs/DISASTER_RECOVERY.md:84]: Loving these docs 😍

File: disaster_recovery/restore/disaster_recovery.py (hunk 1 above)
Comment (jamesrwarren) [reply disaster_recovery/restore/disaster_recovery.py]: Good call

File: disaster_recovery/restore/disaster_recovery.py (hunk 2 above)
Comment (jamesrwarren) [reply disaster_recovery/restore/disaster_recovery.py]: Yeah not sure why i did that. Very naughty.

File: disaster_recovery/restore/disaster_recovery.py (hunk 3 above)
Comment (jamesrwarren) [reply disaster_recovery/restore/disaster_recovery.py]: :)

**Review by acsauk — APPROVED** [review]
:shipit:

**Pull request merged**